pageserver/src/layered_repository/disk_btree.rs

```diff
@@ -513,18 +560,14 @@ where
     }
 
     fn flush_node(&mut self) -> Result<()> {
-        let last = self.stack.pop().unwrap();
+        // Append the downlink to the parent if there is no more items in the stack
+        let last = self.stack.pop_with(|last| BuildNode::new(last.level + 1));
+
         let buf = last.pack();
         let downlink_key = last.first_key();
         let downlink_ptr = self.writer.write_blk(buf)?;
 
-        // Append the downlink to the parent
-        if self.stack.is_empty() {
-            self.stack.push(BuildNode::new(last.level + 1));
-        }
-        self.append_internal(&downlink_key, Value::from_blknum(downlink_ptr))?;
-
-        Ok(())
+        self.append_internal(&downlink_key, Value::from_blknum(downlink_ptr))
     }
```

Comment:
I'm not sure this is really an improvement in readability. I think the biggest problem with the old version was that it was lacking comments. How about this:
```
/// Flush the bottommost node in the stack to disk. Appends a downlink to its parent,
/// and recursively flushes the parent too, if it becomes full. If the root page becomes full,
/// creates a new root page, increasing the height of the tree.
fn flush_node(&mut self) -> Result<()> {
    // Get the current bottommost node in the stack and flush it to disk.
    let last = self.stack.pop().unwrap();
    let blkno = self.writer.write_blk(last.pack())?;

    // Append the downlink to the parent. If there is no parent, ie. this was the root page,
    // create a new root page, increasing the height of the tree.
    if self.stack.is_empty() {
        self.stack.push(BuildNode::new(last.level + 1));
    }
    let downlink_key = last.first_key();
    let downlink = Value::from_blknum(blkno);
    self.append_internal(&downlink_key, downlink)?;

    Ok(())
}
```
And maybe also add a comment to DiskTreeBuilder::stack field above, to explain that it's never empty except during the brief moment above, and after finish() has drained it.

Reply:
I think type system guarantees + comments is a better option than just comments. So the intention was to add guarantees while retaining readability. Here at any `.unwrap()` reader asks himself a question, can it trigger? Why it is impossible? Going back and forth between unwraps/comments requires some cycles. Another point is that code changes can break the invariant and comments won't help with that. IMO NonEmptyVec API clearly indicates the invariant and callers can use it as a building block without worrying about the invariant preservation by themselves

Reply:
Yeah, I see that point too. At the same time, the `pop_ with` function is really funky. It makes me wonder, when is the closure called. Having an explicit `if self.stack.len()` check is more clear IMHO.

Reply:
It's kind of magical how the new node is created here. If the stack becomes empty but you didn't expect that, sure, you avoid the panic. But you unexpectedly initialize it with a new node instead. That could lead to hard-to-debug runtime bugs instead. If the stack is not supposed to be empty, but it is, I think a panic makes a lot of sense. That's exactly the kind of violation of an invariant that panics are meant for.

If you want to have a better notation for representing a stack that cannot be empty, maybe add a `NonEmptyVec` like you did here. But instead of `pop_with`, have a `pop()` that panics if the stack would become empty. With that, the caller could look like this:

```
// If we're flushing the root node, create a new root node above it, to hold the downlink
if self.stack.len() == 1 {
   self.stack.insert(0, BuildNode::new(last.level + 1);
}
last = self.stack.pop();
...
```

Do we have other uses for NonEmptyVec?

Reply:
> It's kind of magical how the new node is created here.

Maybe the name of the function could be better. What do you think about `pop_or_replace_with`? That should make the intention clear. 

Other variants, maybe you'll find one of them more suitable:
pop_fill_empty_with
pop_or_substitute_with

> I think a panic makes a lot of sense.

IMO panic is a bad choice for reporting errors. We either need to use catch_unwind, or it will tear down the whole thread which happened to us many times before. We had quite a few discussions on that with @SomeoneToIgnore.

> have a pop() that panics if the stack would become empty.

That makes it basically the same as it was before. That does not prevent invariant violation and incorrect usage will lead to a panic, the exact thing that NonEmptyVec intends to prevent. The sole purpose of the unusual pop implementation is to maintain the invariant regardless of what the user will do. IMO with the proper name, the intention of the function is clear, so the caller can precisely see what happens and when.

> Do we have other uses for NonEmptyVec?

Sometimes use cases for it pop up, currently, I'm not aware of any other places that can use it. Probably in the future new use cases will appear. 

What do you think about changing only the name of the function and leaving everything else as is?

Reply:
I do not want to insist here. I see that you don't like the approach so it is not a win-win. I still think that abstraction is useful and all that is said about the guarantees is true. I wanted to come to an agreement, but that failed so the obvious steep back is to use just `expect` instead of `unwrap` and add a comment on a `stack` field. I'll create another PR with these changes and try to be better at convincing you next time :)